js/libs/builder.js

```diff
@@ -323,6 +323,44 @@ Fliplet().then(function() {
           Fliplet.Studio.emit('widget-mode', 'wide');
         }
       },
+      getDataSourceColumnValues: function(field) {
```

Comment:
one thing that catches my attention, not only the promises that @tonytlwu mentioned, but also "callback hell". Might be worth breaking it up a bit to make it a little cleaner with awaits.

thoughts @tonytlwu @dwfliplet ?

```
async function getDataSourceColumnValues(field) {
  const id = field.dataSourceId;
  const column = field.column;

  if (!id || !column) {
    return;
  }

  try {
    const cachedValues = await Fliplet.Cache.get({
      key: `${id}-${column}`,
      expire: 60
    });

    if (cachedValues) {
      field.options = cachedValues;
    } else {
      const connection = await Fliplet.DataSources.connect(id);
      const values = await connection.getIndex(column);
      field.options = values.map(option => ({
        id: option ? option.toString().trim() : '',
        label: option ? option.toString().trim() : ''
      }));
      Fliplet.Cache.set({ key: `${id}-${column}`, value: field.options });
    }
  } catch (error) {
    // Handle errors, e.g., log or provide user feedback
    console.error('Error in getDataSourceColumnValues:', error);
  }
}
```